Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PMLPsatdAlgorithm.cpp

```diff
@@ -124,32 +223,87 @@ void PMLPsatdAlgorithm::InitializeSpectralCoefficients (
         Array4<Real> S_ck = S_ck_coef[mfi].array();
+        Array4<Real> C1   = C1_coef  [mfi].array();
+        Array4<Real> C2   = C2_coef  [mfi].array();
+        Array4<Real> C3   = C3_coef  [mfi].array();
+        Array4<Real> C4   = C4_coef  [mfi].array();
+        Array4<Real> C5   = C5_coef  [mfi].array();
+        Array4<Real> C6   = C6_coef  [mfi].array();
+        Array4<Real> C7   = C7_coef  [mfi].array();
+        Array4<Real> C8   = C8_coef  [mfi].array();
+        Array4<Real> C9   = C9_coef  [mfi].array();
+        Array4<Real> C10  = C10_coef [mfi].array();
+        Array4<Real> C11  = C11_coef [mfi].array();
+        Array4<Real> C12  = C12_coef [mfi].array();
+        Array4<Real> C13  = C13_coef [mfi].array();
+        Array4<Real> C14  = C14_coef [mfi].array();
+        Array4<Real> C15  = C15_coef [mfi].array();
+        Array4<Real> C16  = C16_coef [mfi].array();
+        Array4<Real> C17  = C17_coef [mfi].array();
+        Array4<Real> C18  = C18_coef [mfi].array();
+        Array4<Real> C19  = C19_coef [mfi].array();
+        Array4<Real> C20  = C20_coef [mfi].array();
+        Array4<Real> C21  = C21_coef [mfi].array();
+        Array4<Real> C22  = C22_coef [mfi].array();
 
         // Loop over indices within one box
-        ParallelFor(bx,
-        [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept
+        ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept
         {
-            // Calculate norm of vector
-            const Real k_norm = std::sqrt(
-                std::pow(modified_kx[i], 2) +
+            const Real kx = modified_kx[i];
 #if (AMREX_SPACEDIM==3)
-                std::pow(modified_ky[j], 2) +
-                std::pow(modified_kz[k], 2));
+            const Real ky = modified_ky[j];
+            const Real kz = modified_kz[k];
 #else
-                std::pow(modified_kz[j], 2));
+            constexpr Real ky = 0.;
+            const     Real kz = modified_kz[j];
 #endif
 
+            // Calculate norm of vector
+            const Real k_norm = std::sqrt(kx*kx + ky*ky + kz*kz);
+            const Real k2 = k_norm * k_norm;
+
+            const Real kx2 = kx*kx;
+            const Real ky2 = ky*ky;
+            const Real kz2 = kz*kz;
+
             // Calculate coefficients
-            constexpr Real c = PhysConst::c;
-            if (k_norm != 0){
-                C(i,j,k) = std::cos(c*k_norm*dt);
-                S_ck(i,j,k) = std::sin(c*k_norm*dt)/(c*k_norm);
-            } else { // Handle k_norm = 0, by using the analytical limit
-                C(i,j,k) = 1.;
-                S_ck(i,j,k) = dt;
+            constexpr Real c  = PhysConst::c;
+            constexpr Real c2 = c*c;
+
+            // Coefficients for k_norm = 0 do not need to be set
+            if (k_norm != 0.) {
+
+                const Real omega = c * k_norm;
+
+                C(i,j,k)    = std::cos(omega * dt);
+                S_ck(i,j,k) = std::sin(omega * dt) / omega;
+
```

Comment:
It seems that the coefficients `C1` to `C22` can all be computed with simple operation from a small number of coefficients (e.g. `C`, `S_ck`, etc.). Therefore, I would suggest to only save `C` and `S_ck` (and maybe `inv_k2 = 1./k2`) in the class instance, and compute `C1` to `C22` on the fly, inside the field pusher kernel.

This would save memory, esp. on GPU, since these 22 3D arrays might take up quite a bit of space.

Reply:
@RemiLehe Thank you for your suggestion. I implemented the change in https://github.com/ECP-WarpX/WarpX/pull/1513/commits/358e7242a4c09db138f0873490397858191ddbaf.